src/Streamly/FileSystem/Event.hs

```diff
@@ -0,0 +1,239 @@
+module Streamly.FileSystem.Event
+    ( Config
+    , Event
+    , Toggle
+    , setAccessed
+    , defaultConfig
+    , emptyConfig
+    , setAllEvents
+
+    -- ** Events of Interest
+    , setMetadataChanged
+    , setModifiedFileName
+    , setModifiedDirName
+    , setModifiedSize
+    , setModifiedLastWrite
+    , setModifiedSecurity
+
+    -- ** Watch APIs
+    , watch
+
+    -- * Handling Events
+    , getAbsPath
+    , getRelativePath
+    , getRootPath
+
+    -- ** Item CRUD events
+    , isCreated
+    , isDeleted
+    , isMoved
+    , isMovedFrom
+    , isMovedTo
+    , isModified
+
+    -- ** Exception Conditions
+    , isOverflow
+
+    -- * Debugging
+    , showEvent
+    )
+where
+
+import Data.List.NonEmpty (NonEmpty)
+import Data.Word (Word8)
+import Streamly.Prelude (SerialT)
+import Streamly.Internal.Data.Array.Storable.Foreign (Array)
+import System.FilePath ((</>))
+
+#if !defined(CABAL_OS_WINDOWS)
+import qualified Streamly.Internal.Data.Array.Storable.Foreign as Array
+import qualified Streamly.Internal.Data.Stream.IsStream as Stream
+#endif
+#if defined(CABAL_OS_DARWIN)
+import qualified Streamly.Internal.FileSystem.Event.Darwin as Event
+#elif defined(CABAL_OS_LINUX)
+import qualified Streamly.Internal.FileSystem.Event.Linux as Event
+#elif defined(CABAL_OS_WINDOWS)
+import qualified Streamly.Internal.FileSystem.Event.Windows as Event
+#else
+#error "FS Events not supported on this platform
+#endif
+#if !defined(CABAL_OS_WINDOWS)
+import Data.Functor.Identity (runIdentity)
+import qualified Streamly.Internal.Unicode.Stream as U
+#endif
+-------------------------------------------------------------------------------
+-- Utilities
+-------------------------------------------------------------------------------
+
+#if !defined(CABAL_OS_WINDOWS)
+utf8ToString :: Array Word8 -> String
+utf8ToString = runIdentity . Stream.toList . U.decodeUtf8' . Array.toStream
+#endif
+
+--toUtf8 :: MonadIO m => String -> m (Array Word8)
+--toUtf8 = Array.fromStream . Unicode.encodeUtf8' . Stream.fromList
+
+-------------------------------------------------------------------------------
+-- Main
+-------------------------------------------------------------------------------
+type Toggle = Event.Toggle
+type Config = Event.Config
+type Event = Event.Event
+
+setAccessed :: Toggle -> Config -> Config
+#if defined(CABAL_OS_WINDOWS)
+setAccessed = Event.setModifiedLastWrite
+#elif defined(CABAL_OS_LINUX)
+setAccessed = Event.setAccessed
+#elif defined(CABAL_OS_DARWIN)
+setAccessed = undefined
+#endif
+
+setMetadataChanged :: Toggle -> Config -> Config
+#if defined(CABAL_OS_WINDOWS)
+setMetadataChanged = Event.setModifiedAttribute
+#elif defined(CABAL_OS_LINUX)
+setMetadataChanged = Event.setMetadataChanged
```

Comment:
If these two APIs have the same functionality on windows and linux then we should use the same name for them on both platforms in the platform specific module. Same holds for any other such APIs.